File: hadoop/energy_score_plots.m
#!/usr/bin/env octave
# 
# energy_score_plots.m - Plot generating octave script for trident results.
#
# Expects a tab-delimited list of energy-score pairs, with the first three
# columns containing energy, score and frequency, respectively.
#
# Command line arguments: filename and optional title (string in quotation marks).
#
# If the scale_by_two variable is non-zero, the frequency values are divided by two.

scale_by_two = 0;

function retval = extract_data(matrix,column)
  retval = horzcat(matrix(:,column),matrix(:,3));
  retval = reduce_matrix(retval);
  retval = sortrows(retval,1);
endfunction

the_title = "Frequency";
arg_list = argv();
if nargin < 1
  printf("Needs a sorted score file.\n");
  exit(42);
endif

if nargin > 1
  the_title = arg_list{2};
endif

infile = arg_list{1};

data = dlmread(infile,"\t",1,0);

# Scale data?
if scale_by_two
  printf("REDUCED COUNT BY 2!!!!!!!\n");
  if min(data(:,3)) == 1
    printf("MINIMUM VALUE WAS 1!!!!\n");
  endif
  data(:,3) /= 2;
endif

# Energy data
reduced_data = extract_data(data,1);
bar(reduced_data(:,1),reduced_data(:,2));
xlabel("Energy (kcal)");ylabel("Frequency");title(the_title);
saveas(1,"energy.png");

# Score data
reduced_data = extract_data(data,2);
bar(reduced_data(:,1),reduced_data(:,2));
xlabel("Score");ylabel("Frequency");title(the_title);
saveas(1,"score.png");

File: hadoop/energy_score_plots.py
import numpy as np
import matplotlib.pyplot as plt
from getopt import getopt
from sys import argv

short_opts = "s:t:"
long_opts = ["scale=","title="]

def print_usage():
    print("Usage: energy_score_plots.py [-s INT] <input file>")

def reduce_array(arr, sort_col, freq_col):
    unique_keys = np.unique(arr[:,sort_col])
    
    reduced_data = []
    for i in unique_keys:
        new_freq = np.sum(arr[arr[:,sort_col] == i][:,freq_col])
        reduced_data.append([i,new_freq])
    
    return np.array(reduced_data).reshape(-1,2)


# defaults
scale_factor = 1
the_title = "Frequency"

(opts,args) = getopt(argv[1:],short_opts,long_opts)

if len(args) != 1:
    print_usage()
    exit(1)

infilename = args[0]

for (opt,optarg) in opts:
    while opt[0] == '-':
        opt = opt[1:]
    if opt in ["s","scale"]:
        scale_factor = float(optarg)
    elif opt in ["t","title"]:
        the_title = optarg
    else:
        print("Unknown flag: %s" % opt)
        print_usage()
        exit(1)
    
data = np.genfromtxt(infilename, delimiter="\t",skiprows=1)

if scale_factor != 1:
    data[:,2] /= scale_factor
    print("Frequency column (2) was divided by %s" % scale_factor)


# Make plots
for plot_info in [("Energy (kcal)", 0,"energy_test.png"), ("Score", 1,"score_test.png")]:
    reduced_data = reduce_array(data,plot_info[1],2)
    plt.figure()
    plt.bar(reduced_data[:,0],reduced_data[:,1])
    plt.title(the_title)
    plt.xlabel(plot_info[0])
    plt.ylabel("Frequency")
    plt.savefig(plot_info[2])
